src/foam/blob/RestBlobService.java

```diff
@@ -59,23 +59,19 @@ public Blob put_(X x, Blob blob) {
       connection.setRequestProperty("Connection", "keep-alive");
       connection.setRequestProperty("Content-Type", "application/octet-stream");
 
-      //output blob into connection
-      long chunk = 0;
-      long size = blob.getSize();
-      long chunks = (long) Math.ceil((double) size / (double) BUFFER_SIZE);
-      Buffer buffer = new Buffer(BUFFER_SIZE, ByteBuffer.allocate(BUFFER_SIZE));
+      // get connection ouput stream
       os = connection.getOutputStream();
 
+      //output blob into connection
+      int chunk = 0;
+      int size = blob.getSize();
+      int chunks = (int) Math.ceil((double) size / (double) BUFFER_SIZE);
+
       while ( chunk < chunks ) {
-        buffer = blob.read(buffer, chunkOffset(chunk));
-        byte[] buf = buffer.getData().array();
-        os.write(buf, 0, (int) buffer.getLength());
-        buffer.getData().clear();
+        blob.read(os, chunk * BUFFER_SIZE, size);
```

Comment:
"size" is the total size of the blob, but we're trying to read it in chunks of size BUFFER_SIZE.

The third parameter should be ( size % BUFFER_SIZE == 0 ? BUFFER_SIZE : size % BUFFER_SIZE ) or similar.

If size % BUFFER_SIZE is 0, then all chunks are the same size, if non-zero then it should give you the size of the final chunk.

Reply:
I think this can actually just be simplified to ```blob.read(os, 0, size);```. The read function can perform it's own internal chunking.

Reply:
I agree